### PR TITLE
[AIE2] Legalize V2S32 G_LOAD.

### PR DESCRIPTION
**Pull request by SagarMaheshwari99**
A more suitable approach will be to just return false for TLI.allowsMemoryAccess() for v2s32 G_LOAD since AIE2 doesn't allow it. But just a placeholder for 5 of our failing tests. https://jira.xilinx.com/browse/AIECC-667

### CODE REVIEW

File: llvm/test/CodeGen/AIE/aie2/GlobalISel/legalize-64bit-load-store.mir [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+body:             |
+  bb.1.entry:
+    liveins: $p0
+    ; CHECK-LABEL: name: load_v2s32
+    ; CHECK: liveins: $p0
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(p0) = COPY $p0
+    ; CHECK-NEXT: [[LOAD:%[0-9]+]]:_(s32) = G_LOAD [[COPY]](p0) :: (load (s32), align 8)
+    ; CHECK-NEXT: [[C:%[0-9]+]]:_(s20) = G_CONSTANT i20 4
+    ; CHECK-NEXT: [[PTR_ADD:%[0-9]+]]:_(p0) = G_PTR_ADD [[COPY]], [[C]](s20)
```
Comment (krishnamtibrewala) [inline llvm/test/CodeGen/AIE/aie2/GlobalISel/legalize-64bit-load-store.mir]: Does `combine_load_store_increment`  combiner pass combines these Load/Store into post or pre increment load/store ?

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp [hunk 2]
```diff
@@ -388,6 +388,17 @@ AIELegalizerInfo::AIELegalizerInfo(const AIEBaseSubtarget &ST) {
...
       .lowerIfMemSizeNotPow2()
+      .bitcastIf(
+          [=](const LegalityQuery &Query) {
+            const LLT &Ty = Query.Types[0];
+            return Ty.isVector() && Ty.getSizeInBits() == 64;
+          },
+          [=](const LegalityQuery &Query) {
+            const LLT Ty = Query.Types[0];
+            unsigned Size = Ty.getSizeInBits();
+            assert(Size % 64 == 0);
```
Comment (krishnamtibrewala) [inline llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: Please add a string with assert, helps in debugging.

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp [hunk 3]
```diff
@@ -388,6 +388,17 @@ AIELegalizerInfo::AIELegalizerInfo(const AIEBaseSubtarget &ST) {
       })
       .widenScalarToNextPow2(0)
       .lowerIfMemSizeNotPow2()
+      .bitcastIf(
+          [=](const LegalityQuery &Query) {
+            const LLT &Ty = Query.Types[0];
+            return Ty.isVector() && Ty.getSizeInBits() == 64;
```
Comment (konstantinschwarz) [inline llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: While you are here, maybe also support vector sizes == 32?
Bitcasting such vectors to scalars yields much better code:

```
void put(v4int8 *p, v4int8 val) {    *p = val; }
```

with current legalization strategy (i.e. `.lower()`):
```
    liveins: $p0, $r0
  
    %0:_(p0) = COPY $p0
    %1:_(<4 x s8>) = COPY $r0
    %16:_(s32) = G_BITCAST %1(<4 x s8>)
    %17:_(s32) = G_CONSTANT i32 8
    %18:_(s32) = G_LSHR %16, %17(s32)
    %19:_(s32) = G_CONSTANT i32 16
    %20:_(s32) = G_LSHR %16, %19(s32)
    %21:_(s32) = G_CONSTANT i32 24
    %22:_(s32) = G_LSHR %16, %21(s32)
    G_STORE %16(s32), %0(p0) :: (store (s8) into %ir.p, align 4, !tbaa !2)
    %7:_(s20) = G_CONSTANT i20 1
    %6:_(p0) = G_PTR_ADD %0, %7(s20)
    G_STORE %18(s32), %6(p0) :: (store (s8) into %ir.p + 1, basealign 4, !tbaa !2)
    %9:_(s20) = G_CONSTANT i20 2
    %8:_(p0) = G_PTR_ADD %0, %9(s20)
    G_STORE %20(s32), %8(p0) :: (store (s8) into %ir.p + 2, align 2, basealign 4, !tbaa !2)
    %11:_(s20) = G_CONSTANT i20 3
    %10:_(p0) = G_PTR_ADD %0, %11(s20)
    G_STORE %22(s32), %10(p0) :: (store (s8) into %ir.p + 3, basealign 4, !tbaa !2)
    PseudoRET implicit $lr
```

Bitcasting to `s32`:
```
    %0:_(p0) = COPY $p0
    %1:_(<4 x s8>) = COPY $r0
    %2:_(s32) = G_BITCAST %1(<4 x s8>)
    G_STORE %2(s32), %0(p0) :: (store (s32) into %ir.p, !tbaa !2)
    PseudoRET implicit $lr
```

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp (hunk 2 above)
Comment (konstantinschwarz) [reply llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: Agreed. Also, just use `Size` in `LLT:scalar(Size)` below.
That also prevents an unused variable warning in non-assert builds.

File: llvm/test/CodeGen/AIE/aie2/GlobalISel/legalize-64bit-load-store.mir (hunk 1 above)
Comment (konstantinschwarz) [reply llvm/test/CodeGen/AIE/aie2/GlobalISel/legalize-64bit-load-store.mir]: yes

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp (hunk 2 above)
Comment (martien-de-jong) [reply llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: TBH, I think assert generates a perfectly readable string for this condition. Does it really need to be cluttered with "type size is not a multiple of 64 bits", effectively saying the same?

Comment (SagarMaheshwari99) [reply llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: I agree with @martien-de-jong. I will make the "Size" change though..

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp (hunk 3 above)
Comment (SagarMaheshwari99) [reply llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: added.

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp [hunk 4]
```diff
@@ -389,6 +389,18 @@ AIELegalizerInfo::AIELegalizerInfo(const AIEBaseSubtarget &ST) {
...
+      .bitcastIf(
+          [=](const LegalityQuery &Query) {
+            const LLT &Ty = Query.Types[0];
+            return Ty.isVector() &&
+                   (Ty.getSizeInBits() == 64 || Ty.getSizeInBits() == 32);
+          },
+          [=](const LegalityQuery &Query) {
+            const LLT Ty = Query.Types[0];
+            unsigned Size = Ty.getSizeInBits();
+            assert(Size % 64 == 0 || Size % 32 == 0);
```
Comment (martien-de-jong) [inline llvm/lib/Target/AIE/AIELegalizerInfo.cpp]: nit: I think this can be simplified to `Size % 32 == 0`

**Review by martien-de-jong — DISMISSED** [review]

File: llvm/lib/Target/AIE/AIELegalizerInfo.cpp [hunk 5]
```diff
@@ -389,6 +389,18 @@ AIELegalizerInfo::AIELegalizerInfo(const AIEBaseSubtarget &ST) {
...
       .widenScalarToNextPow2(0)
       .lowerIfMemSizeNotPow2()
+      .bitcastIf(
+          [=](const LegalityQuery &Query) {
+            const LLT &Ty = Query.Types[0];
+            return Ty.isVector() &&
+                   (Ty.getSizeInBits() == 64 || Ty.getSizeInBits() == 32);
+          },
+          [=](const LegalityQuery &Query) {
+            const LLT Ty = Query.Types[0];
```
Comment (andcarminati) [inline llvm/lib/Target/AIE/AIELegalizerInfo.cpp:399]: nit: `const unsigned Size`

**Review by andcarminati — APPROVED** [review]
LGTM.

**Pull request merged**